File: frontend-react/src/api/notifications.ts
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import api from "./client";

export interface NotificationItem {
  id: string;
  run_id: string;
  target_role: string;
  title: string;
  message: string;
  link: string | null;
  is_read: boolean;
  created_at: string;
}

export function useNotifications(role: string) {
  return useQuery<NotificationItem[]>({
    queryKey: ["notifications", role],
    queryFn: () =>
      api.get(`/api/notifications?role=${role}`).then((r) => r.data),
    refetchInterval: 10000,
  });
}

export function useUnreadCount(role: string) {
  return useQuery<NotificationItem[]>({
    queryKey: ["notifications-unread", role],
    queryFn: () =>
      api.get(`/api/notifications?role=${role}&unread=true`).then((r) => r.data),
    refetchInterval: 10000,
    select: (data) => data,
  });
}

export function useMarkRead() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (notificationId: string) =>
      api.post(`/api/notifications/${notificationId}/read`).then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["notifications"] });
      qc.invalidateQueries({ queryKey: ["notifications-unread"] });
    },
  });
}

export function useMarkAllRead() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (role: string) =>
      api.post(`/api/notifications/read-all?role=${role}`).then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["notifications"] });
      qc.invalidateQueries({ queryKey: ["notifications-unread"] });
    },
  });
}

File: app/api/notifications.py
from __future__ import annotations

from uuid import UUID
from typing import Optional, List
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.services import notification_service

router = APIRouter(prefix="/api/notifications", tags=["notifications"])


class NotificationResponse(BaseModel):
    id: str
    run_id: str
    target_role: str
    title: str
    message: str
    link: Optional[str] = None
    is_read: bool
    created_at: datetime

    class Config:
        from_attributes = True


@router.get("", response_model=List[NotificationResponse])
async def list_notifications(
    role: str = Query(..., description="Role to fetch notifications for"),
    unread: bool = Query(False, description="Only unread"),
    db: AsyncSession = Depends(get_db),
):
    notifs = await notification_service.get_notifications(db, role, unread_only=unread)
    return [
        NotificationResponse(
            id=str(n.id),
            run_id=str(n.run_id),
            target_role=n.target_role,
            title=n.title,
            message=n.message,
            link=n.link,
            is_read=n.is_read,
            created_at=n.created_at,
        )
        for n in notifs
    ]


@router.post("/{notification_id}/read")
async def read_notification(
    notification_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    ok = await notification_service.mark_read(db, notification_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Notification not found")
    return {"status": "read"}


@router.post("/read-all")
async def read_all_notifications(
    role: str = Query(..., description="Role to mark all read for"),
    db: AsyncSession = Depends(get_db),
):
    count = await notification_service.mark_all_read(db, role)
    return {"status": "ok", "marked_read": count}
